### PR TITLE
Use llama precompiled source

### PR DESCRIPTION
**Pull request by mcsgroi**
### Description
This change just modifies the Docker build for the API to use the precompiled llama Docker image as opposed to compiling from source every build. 

### Changes
- Alter API Docker build to remove the need to compile llama from source

### Issue
https://github.com/nsarrazin/serge/issues/48

### CODE REVIEW

**Comment by nsarrazin** [comment]
Hey apologies the `Dockerfile.api` is deprecated and is no longer used. The `Dockerfile` file at the root of the project is the one that is being used. But the same code should work there too! If you feel like moving it to the other file then I'll have a look, sounds like a great fix to solve a lot of headaches haha.

I'll also remove the deprecated dockerfiles, sorry again for forgetting to remove them.

**Comment by mcsgroi** [comment]
@nsarrazin no worries. I believe this should be addressed now. Let me know if you see any issues.

**Comment by nsarrazin** [comment]
Trying to run it from within Serge I get a server error.

When running the `llama` executable directly this is what I get.

```
root@74571ae41425:/usr/src/app# llama -m weights/ggml-alpaca-7B-q4_0.bin
main: seed = 1679773025
llama_model_load: loading model from 'weights/ggml-alpaca-7B-q4_0.bin' - please wait ...
llama_model_load: n_vocab = 32000
llama_model_load: n_ctx   = 512
llama_model_load: n_embd  = 4096
llama_model_load: n_mult  = 256
llama_model_load: n_head  = 32
llama_model_load: n_layer = 32
llama_model_load: n_rot   = 128
llama_model_load: f16     = 2
llama_model_load: n_ff    = 11008
llama_model_load: n_parts = 1
llama_model_load: type    = 1
Illegal instruction (core dumped)
```

Is this something you know about ? 

I'm running this on:
* **CPU**: Ryzen 5 5600X (avx & avx2 compatible) 
* **OS**: Pop Os 22.04, kernel 6.2.0-76060200-generic
* **Docker Compose** version v2.16.0

**Comment by mcsgroi** [comment]
Interesting. I wasn't seeing that before, but I just did a complete rebuild on the image and am seeing it now. 

I'll have to do some digging. No ideas offhand as to why this might be happening.

**Comment by nsarrazin** [comment]
I'm also trying to run the docker command  directly from the README of llama.cpp with

 `docker run -v .:/models ghcr.io/ggerganov/llama.cpp:light -m models/ggml-alpaca-7B-q4_0.bin -p "Building a website can be done in 10 simple steps:" -n 512`  

and no success so far there either. 

So I'm not sure if it's a bug related to serge or the docker image itself

**Comment by mcsgroi** [comment]
I've tried a few different versions of the Docker copied binary in Serge now and all of them are exhibiting the same behavior.

I'm hesitant to say that llama.cpp Docker image has a problem. If you look through this file, it more or less does exactly what your source build does https://github.com/ggerganov/llama.cpp/blob/master/.devops/main.Dockerfile

It looks like you have the git pull for the llama source pinned to a specific commit. Let me see if I can find a release for that commit to test in Docker.

**Comment by mcsgroi** [comment]
Looks like the Docker image tied with the commit pinned in the current Serge Dockerfile is still exhibiting the same behavior. Still trying to wrap my head around how the Docker image binary could possibly be different from the one being compiled from source in Serge if they execute the same steps.

Have you by chance tested the current version of Serge without this change to see if this issue is present? I would test myself, but I'm having trouble with the source compilation step on my machine.

**Comment by nsarrazin** [comment]
Alright current findings:

* building llama in serge using the current dockerfile: works fine I get an output
* building llama in serge, updating the branch used to `master-c2b25b6` (latest as of testing) works fine too, no issues

* replacing the build step with `FROM ghcr.io/ggerganov/llama.cpp:light as llama_builder`, I get the error `Illegal instruction (core dumped)`

* running the llama.cpp image directly, bypassing serge completely with `
docker run -v .:/models ghcr.io/ggerganov/llama.cpp:light -m models/ggml-alpaca-7B-q4_0.bin -p "Building a website can be done in 10 simple steps:" -n 512
` I also get an error.

**Comment by nsarrazin** [comment]
might be worth trying with a new branch in the serge dockerfile like `master-c2b25b6` if it improves your compilation errors.

**Comment by gaby** [comment]
What happens if you run `ghcr.io/ggerganov/llama.cpp:light` alone?

**Comment by nsarrazin** [comment]
> What happens if you run `ghcr.io/ggerganov/llama.cpp:light` alone?

That's what I did with the last step in my previous message, the command is taken straight from the readme, except the bind mount which I changed. 

Is it working for you ?

**Comment by gaby** [comment]
I downloaded the model, but it complains that it's too old... Which one are you using?

**Comment by gaby** [comment]
This is the one I used: https://huggingface.co/Pi3141/alpaca-7B-ggml/tree/main

**Comment by nsarrazin** [comment]
It's too old you gotta convert it. Serge does it for you automatically. 

The script is here: https://github.com/nsarrazin/serge/blob/main/api/utils/convert.py

**Comment by mcsgroi** [comment]
Just found this issue: https://github.com/ggerganov/llama.cpp/issues/402

Seems pretty relevant here based on the behavior. 

Same as before, I have no idea how this correlates to using a precompiled binary vs compiling it from scratch though.

**Comment by gaby** [comment]
@mcsgroi So it seems the image gets build on every commit to master, maybe there was an issue. Try again with the latest one.

**Comment by mcsgroi** [comment]
That does appear to be the case. I've tried a few images all the way back to the commit that was pinned in this original build process. All seemed to yield the same error result.

However, I just pulled the latest image from a few minutes ago and, wouldn't you it, it worked. I'll get this updated.

**Comment by mcsgroi** [comment]
Updated. Would appreciate if someone would give this a run on their machine to sanity check this version works as expected.

**Comment by nsarrazin** [comment]
> Updated. Would appreciate if someone would give this a run on their machine to sanity check this version works as expected.

On it!

File: Dockerfile
Comment (gaby) [inline Dockerfile:3]: You can probably remove this line. In the 2nd stage just copy the `main` binary into `/usr/bin/llama`

**Comment by nsarrazin** [comment]
Ha incredible, it seems to work now ? Let's make sure the version is frozen and let's just use this one for now :laughing:

**Review by gaby — APPROVED** [review]
LGTM

File: Dockerfile
Comment (nsarrazin) [reply Dockerfile:3]: @mcsgroi if you could please do this change while we're here in the PR, i'll approve & merge afterwards

Comment (mcsgroi) [reply Dockerfile]: Addressed.

**Comment by nsarrazin** [comment]
I asked the folks on discord to have a review if possible. So I'll give them some time, but I probably will merge soon. Works great for me! Thanks, I think it's a real upgrade.

**Comment by nsarrazin** [comment]
Someone with a Mac M1 got the following issue: 

```
 [+] Building 0.4s (4/4) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                          0.0s
 => => transferring dockerfile: 32B                                                                                                                                                                                           0.0s
 => [internal] load .dockerignore                                                                                                                                                                                             0.0s
 => => transferring context: 32B                                                                                                                                                                                              0.0s
 => [internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                                                               0.3s
 => ERROR [internal] load metadata for ghcr.io/ggerganov/llama.cpp:light-19726169b379bebc96189673a19b89ab1d307659                                                                                                             0.3s
------
> [internal] load metadata for ghcr.io/ggerganov/llama.cpp:light-19726169b379bebc96189673a19b89ab1d307659:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: no match for platform in manifest sha256:cf18d117cbf29013c5484d87e27c5dd478bc69cd448ef8b937d2847a7c1f8b81: not found
```

**Comment by nsarrazin** [comment]
Maybe we could try to use a fallback somehow  ? If it fails to find an image, compile manually.

**Comment by willjasen** [comment]
Heya! The website is spinning up but when trying to chat, I get this on two different machines with 16 GB+ of RAM.

image

**Comment by mcsgroi** [comment]
@nsarrazin I'm not sure I have a good answer for your request. I'd have to do some research, but I don't know of a simple way to conditionally import different Docker images. I noticed that @gaby introduced a related change for this. 

If there's something specific you had in mind and a change is still required, let me know.

**Comment by mcsgroi** [comment]
@willjasen that looks like the issue I was seeing before I updated the image referenced in this PR (commit: [530f2b6](https://github.com/nsarrazin/serge/pull/66/commits/530f2b68422c759892244f445fd217c8327e8c34)). Have you tried rebuilding since that update `docker build up -d --build`?

**Comment by willjasen** [comment]
> @willjasen that looks like the issue I was seeing before I updated the image referenced in this PR (commit: [530f2b6](https://github.com/nsarrazin/serge/pull/66/commits/530f2b68422c759892244f445fd217c8327e8c34)). Have you tried rebuilding since that update `docker build up -d --build`?

I tried "docker build up -d --build" and it's the same thing.

**Comment by nsarrazin** [comment]
> Heya! The website is spinning up but when trying to chat, I get this on two different machines with 16 GB+ of RAM.
> 
> image

Hey! Can you tell me more about your hardware ? OS/CPU etc.

And also if you could run `docker compose exec serge llama -m weights/your_model_goes_here.bin` and tell me the output, would be great.

**Comment by gaby** [comment]
@mcsgroi Theres been quite a few commits, can you update the sha256 to the latest one?

**Comment by nsarrazin** [comment]
I think we need https://github.com/ggerganov/llama.cpp/pull/514 to be merged before merging this right ? Otherwise we lose ARM support.

**Comment by gaby** [comment]
You can conditionally use docker tags, but you can't lets say "Try the precompile if it fails build from source". That's not the use-case of the Docker.

**Comment by gaby** [comment]
> I think we need [ggerganov/llama.cpp#514](https://github.com/ggerganov/llama.cpp/pull/514) to be merged before merging this right ? Otherwise we lose ARM support.

Correct, it has 1 approval l, waiting on the 2nd one 😁

**Comment by gaby** [comment]
@mcsgroi Can you update the tag to this one:

https://github.com/ggerganov/llama.cpp/pkgs/container/llama.cpp/80431708?tag=light-8c2ec5e21d580c99e257c3cfddcf21fa53229aa4

It has support for linux/arm64

**Comment by mcsgroi** [comment]
Has anyone tested that new sha to make sure it continues to function as expected?

**Comment by mcsgroi** [comment]
Just tested myself and found that the version suggested by @gaby and the version updated by @nsarrazin exhibit the same bug we found earlier in this PR thread. The only version I've seen work is `light-19726169b379bebc96189673a19b89ab1d307659`

**Comment by nsarrazin** [comment]
So recent findings. I tried the version mentioned by @gaby :

If I use the image hosted on the llama.cpp repo, it fails to load. `Illegal instruction  Illegal instruction (core dumped)`

But if I use the exact same commit, but I build it from scratch it works without issues. Really confused by this behavior.

The version mentioned by @mcsgroi is the only one that works but it doesn't have ARM compatibility.

Seems like we're kind of stuck, am I missing something ?

**Comment by mcsgroi** [comment]
My best guess is that there is a build issue in the llama.cpp CI system. I can't think of any other reasons that would happen.

Unless we happen to find a new version of llama.cpp that happened to compile in a way that seems to run I don't see a path forward for this PR. Considering that llama.cpp seems to release incredibly frequently though, it might be worth waiting for a few versions to get released to see if any of them work since it seems to be fairly arbitrary.

**Comment by nsarrazin** [comment]
Maybe we can raise the issue with them to see if they know anything about it.

**Comment by nsarrazin** [comment]
The problem is that this is also blocking the proper use of prebuilt images for serge,  since we're also building `llama.cpp` in the serge image, which has the same error. 

The only way to get around it is to postpone the compiling to the entrypoint of the dockerfile, so it runs on the CPU of the user. This is a whole other issue though, and has a lot of downsides too. I'm not sure what to do.

**Comment by gaby** [comment]
This may be related to what we are seeing on this PR:

https://github.com/ggerganov/llama.cpp/issues/537

**Comment by willjasen** [comment]
[nsarrazin](https://github.com/nsarrazin)
> Hey! Can you tell me more about your hardware ? OS/CPU etc.

I'm running two Ubuntu 22.04 instances, one is a virtual machine and one is bare metal. Both have 4 cores for CPU and at least 16 GB of RAM. I'm only concerned with getting 7B up and running for the moment as I assume the others will follow with correct hardware parameters.

> And also if you could run `docker compose exec serge llama -m weights/your_model_goes_here.bin` and tell me the output, would be great.

I get the same thing as what the chat outputs:
image

**Comment by elbowdonkey** [comment]
I ran into the `always_inline` [error](url) as well (both with this repo, and llama.cpp).

`docker compose up -d` completed without errors. I'm currently waiting for `docker compose exec serge python3 /usr/src/app/api/utils/download.py tokenizer 7B` to complete. I'll report the results here when it's complete.

System Info:

* Xeon E5-2430 0 @ 2.20GHz (x2)
* 384GB Ram
* Freshly installed Ubuntu 22.04.2 LTS

**Comment by gaby** [comment]
I don't think this will work until https://github.com/ggerganov/llama.cpp/discussions/535 is resolved. Since everyone has a different CPU, and the instruction sets will differ.

**Comment by elbowdonkey** [comment]
It appears that `docker compose up -d` and `docker compose exec serge python3 /usr/src/app/api/utils/download.py tokenizer 7B` succeeded, but I have to confess: I don't know how to start the server. Is the `README` missing an instruction? Is `download.py` supposed to also start the web server?

**Comment by elbowdonkey** [comment]
I spoke too soon - I had a firewall rule blocking port 8008, but now with it unblocked I'm able to get to the server.

That said, I seeing errors as responses:

```
main: seed = 1679872256 
llama_model_load: loading model from '/usr/src/app/weights/ggml-alpaca-7B-q4_0.bin' - please wait ... 
llama_model_load: n_vocab = 32000 
llama_model_load: n_ctx = 512 
llama_model_load: n_embd = 4096 
llama_model_load: n_mult = 256 
llama_model_load: n_head = 32 
llama_model_load: n_layer = 32 
llama_model_load: n_rot = 128 
llama_model_load: f16 = 2 
llama_model_load: n_ff = 11008 
llama_model_load: n_parts = 1 
llama_model_load: type = 1
```

**Comment by mcsgroi** [comment]
I second @gaby 's opinion. @nsarrazin what would you like me to do here. I still contend that this is the better workflow pattern, but it's clear from the issue in llama.cpp that this will not work, at least not until changes are made on their end. I'm fine with leaving this open or closing it in the interim. Thoughts?

**Comment by nsarrazin** [comment]
Yeah I also think this will need to be fixed on their end, which we can't wait on. Let's close this and I'll open a relevant issue

**Comment by nsarrazin** [comment]
@mcsgroi @gaby  I opened ticket #82 to continue the discussion there.

**Pull request closed without merging**
